### DISCUSSION TITLE
How to handle `GET method` with body param in CB-TB API

### DISCUSSION THREAD

**Discussion by seokho-son** [cloud-barista/cb-tumblebug#502] [General]
**What happened**
: 
다음 GET API들은 Body로 GET에 대한 조건을 입력 받습니다.
GET ​/ns​/{nsId}​/resources​/filterSpecs
GET ​/ns​/{nsId}​/resources​/filterSpecsByRange

GET ​/lookupImage <- connectionConfig 만
GET ​/lookupImages <- connectionConfig 만
GET ​/lookupSpec <- connectionConfig 만
GET ​/lookupSpecs <- connectionConfig 만

Swagger UI 에서 Body를 포함한 Get을 금지하고 있어서, 다음과 같은 애러가 발생합니다.

Code | Details
-- | --
Undocumented | TypeError: Failed to execute 'fetch' on 'Window': Request with GET/HEAD method cannot have body.

**What you expected to happen**
: 애러 없이 정상 동작

**How to reproduce it (as minimally and precisely as possible)**
: 항시 발생

**Anything else we need to know?**
: 

**Environment**
- Source version or branch: master
- OS: ubuntu
- Others: 

**Proposed solution**
: @jihoon-seo 의 조언에 따라

**Any other context**
:

**Comment by seokho-son** [comment]
PTAL @jihoon-seo

**Comment by jihoon-seo** [comment]
[찾아 보니](https://www.thecodebuzz.com/failed-execute-fetch-on-window-request-get-cannot-have-body/),
[RFC 7231](https://tools.ietf.org/html/rfc7231#section-4.3.1) 에 따르면
> GET is the primary mechanism of information retrieval and the focus of almost all performance optimizations.
> 
> A payload within a GET request message has no defined semantics; sending a payload body on a GET request might cause some existing implementations to reject the request.

라고 합니다.

Resty 도
```Go
client := resty.New()
client.SetAllowGetMethodPayload(true)
```
코드를 실행하여 GET 에 body 를 넣을 수는 있지만
기본적으로는 GET 에 body 를 넣을 수 없습니다.

Swagger, Resty 등의 사례와 RFC 7231을 참고하여
TB REST API 중 JSON body 를 받는 것은
method 를 GET 외의 다른 것으로 변경해야 하지 않을까 싶습니다.

다른 링크:
- https://stackoverflow.com/questions/48749252/swagger-typeerror-failed-to-execute-fetch-on-window-request-with-get-head
  - 두 번째 답변에서는 `[FromQuery]` 를 추가하는 해법 (우회법?) 을 소개하고 있습니다.
  - 동일한 해법에 대한 글: https://stackoverflow.com/questions/57760770/adding-fromquery-to-api-method-solves-swagger-issue
- https://stackoverflow.com/questions/56655554/failed-to-execute-fetch-on-window-request-with-get-head-method-cannot-have
- https://stackoverflow.com/questions/50128551/unhandled-rejection-typeerror-failed-to-execute-fetch-on-window-request
- ...

---

기타:

Spider 초기 코드에서는
`connectionName` 을 Query param 으로 받았었다가,
Tumblebug 팀의 제안을 통해 현재는 JSON body 로 받고 있습니다.

Tumblebug 은 `connectionName` 을 초기부터 JSON body 로 받았는지
아니면 Query param 으로 받았던 때도 있었는지 기억이 잘 나지 않습니다.

**Comment by seokho-son** [comment]
@jihoon-seo 
그럼~ 아래 네 가지 방법이 있는건가욤

- Get with Query Param (Query URL이 다소 복잡)
- Post with JSON Body (조회인데 Post라 마음이 불편함)
- Get with JSON (기존 방식, Forced, Swagger 에서는 정상 동작하지 않음)
- Get with Path Param (리소스 Path에 해당하지 않는 경우 있음. 다른 오류도 발생 가능)

**Comment by seokho-son** [comment]
Post with JSON Body (조회인데 Post라 마음이 불편함)
로 일단 PR 올렸습니다.
